File: Algorithms/p114_Flatten_Binary_Tree_to_Linked_List.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        flattenDFS(root);
    }
    TreeNode* flattenDFS(TreeNode* root) {
        if(!root) return 0;
        TreeNode* l_tail = flattenDFS(root->left);
        TreeNode* r_tail = flattenDFS(root->right);
        if(root->left){
            l_tail->right = root->right;
            root->right = root->left;
            root->left = 0;
        }
        return r_tail ? r_tail : l_tail ? l_tail : root;
    }
};

File: Algorithms/p114_Flatten_Binary_Tree_to_Linked_List/p114_Flatten_Binary_Tree_to_Linked_List.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        self.flattenDFS(root)

    def flattenDFS(self, root):
        if not root:
            return None
        l_tail = self.flattenDFS(root.left)
        r_tail = self.flattenDFS(root.right)
        if root.left:
            l_tail.right = root.right
            root.right = root.left
            root.left = None
        return r_tail if r_tail else l_tail if l_tail else root
